_source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md

```diff
@@ -0,0 +1,343 @@
+
+### Create a self-signed certificate
+
+To set up TLS, you'll need a certificate. In real life, you'd buy one from a reputable provider, but for this walkthrough, a self-signed certificate will do the job.
+
+Generate a certificate:
+
+```bash
+openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -days 365
+```
+
+To use the certificate with Kestrel (the ASP.NET Core webserver), we need to create a `pfx` file that has both the public and private keys. You can do that using:
+
+```bash
+openssl pkcs12 -export -out webserver.pfx -inkey key.pem -in cert.pem
+```
+
+As part of the certificate creation process, you'll be prompted to create an Export password. Be sure to take note of this as you'll need it later to use the certificate. You'll also be walked through the configuration process (_Country Name_, _State or Province_, etc.).
+
+### Add the certificate to the Docker image
+
+You've created a certificate in your local filesystem. To use it in your Docker container, modify the Dockerfile to copy it into the final image.
+
+Change the `final' stage to:
+
+```bash
+FROM base AS final
+ENV ASPNETCORE_URLS="https://+"
+ENV ASPNETCORE_Kestrel__Certificates__Default__Path="./webserver.pfx"
+
+WORKDIR /app
+COPY --from=publish /app .
+COPY webserver.pfx .
+ENTRYPOINT ["dotnet", "OktaMvcLogin.dll"]
+```
+
+Setting the ASPNETCORE_URLS environment variable to `https://+` ensures that the webserver only listens for https requests.
+
+Since you've changed the Dockerfile, you can rebuild the image and run it:
+
+```bash
+docker build -t oktamvclogin .
+docker run -e ASPNETCORE_Okta__ClientId="{yourClientId}" \
+-e ASPNETCORE_Okta__ClientSecret="{yourClientSecret}" \
+-e ASPNETCORE_Okta__Domain="{yourOktaDomain}" \
+-e ASPNETCORE_Kestrel__Certificates__Default__Password="{yourExportPassword}" \
+-p 5001:443 \
+oktamvclogin
+```
+
+Notice the additional environment variable with the certificate export password and also, that the port mapping has changed from port 80 to port 443.
+
+You can now navigate to `https://localhost:5001`, and this time, you'll be able to log in and use the sample application correctly.
+
+>Note: Since you're using a self-signed certificate, your browser may display a warning page. You can safely ignore this warning.
+
+## Not the Best Way to Start Your Docker Container
+
+Converting an application to run in Docker is relatively straightforward and offers significant benefits. However, passing a whole load of configuration to `docker run` isn't particularly user-friendly and is error-prone. Thankfully, the good folks at Docker have already come up with a solution to this problem - `docker-compose`.
+
+Using Docker Compose is pretty straightforward. Create a file named `docker-compose.yml` in the same folder as your `Dockerfile` and source code and add the following:
+
+```yaml
```

Comment:
This YAML is incorrect. It needs to be formatted properly.

Reply:
How should it be formatted?

Reply:
Here's a guess based on other Docker compile files I've seen:

```yaml
version: "3"
services:
  web:
    build: .
    image: oktamvclogin
    ports:
      - "5001:443"
    environment:
      - ASPNETCORE_Okta__ClientId: "{yourClientId}-"
      - ASPNETCORE_Okta__ClientSecret: "{yourClientSecret}"
      - ASPNETCORE_Okta__Domain: "{yourOktaDomain}"
      - ASPNETCORE_Kestrel__Certificates__Default__Password: "{yourExportPassword}"
```